File: src/main/java/uk/ac/ebi/tsc/tesk/k8s/convert/data/Job.java
package uk.ac.ebi.tsc.tesk.k8s.convert.data;

import io.kubernetes.client.models.V1Job;
import io.kubernetes.client.models.V1Pod;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Ania Niewielska <aniewielska@ebi.ac.uk>
 * <p>
 * A container for a single Kubernetes job object (can be both a taskmaster and an executor)
 * and its list of worker pods (Kubernetes Pod objects)
 */
public class Job {
    /**
     * Kubernetes job object
     */
    private V1Job job;

    /**
     * The list of Kubernetes pod objects created by the job
     */
    private List<V1Pod> pods;

    public Job(V1Job job) {
        this.job = job;
    }

    public V1Job getJob() {
        return this.job;
    }

    /**
     * Adds a single pod to a list (without any duplication checks)
     */
    public void addPod(V1Pod pod) {
        if (this.pods == null) {
            this.pods = new ArrayList<>();
        }
        this.pods.add(pod);
    }

    public boolean hasPods() {
        return this.pods != null && this.pods.size() > 0;
    }

    /**
     * Returns arbitrarily chosen pod from the list (currently the first one added)
     * or null, if job has no pods.
     */
    public V1Pod getFirstPod() {
        if (!hasPods()) return null;
        return this.pods.get(0);
    }

    /**
     * Returns the list of job pods in the order of addition to the list
     * or empty list, if no pods.
     */
    public List<V1Pod> getPods() {
        if (!hasPods()) return new ArrayList<>();
        return this.pods;
    }
    public void changeJobName(String newName) {
        this.job.getMetadata().name(newName);
        this.job.getSpec().getTemplate().getMetadata().name(newName);
        this.job.getSpec().getTemplate().getSpec().getContainers().get(0).setName(newName);
    }
    public String getJobName() {
        return this.job.getMetadata().getName();
    }

}

File: tesk/api/kubernetes/convert/data/job.py
"""A container for a single Kubernetes job object.

Can be both a taskmaster and an executor, it list of worker pods (Kubernetes
Pod objects).
"""

from enum import Enum
from typing import List, Optional

from kubernetes.client import V1Job, V1ObjectMeta, V1Pod


class Job:
    """Class to list worker pods (Kubernetes Pod objects)."""

    def __init__(self, job: V1Job):
        """Initializes the Job with a Kubernetes job object."""
        self.job: V1Job = job
        self.pods: List[V1Pod] = []

    def get_job(self) -> V1Job:
        """Returns the Kubernetes job object."""
        return self.job

    def add_pod(self, pod: V1Pod):
        """Adds a single pod to the list."""
        self.pods.append(pod)

    def has_pods(self) -> bool:
        """Checks if the job has any pods."""
        return bool(self.pods)

    def get_first_pod(self) -> Optional[V1Pod]:
        """Returns arbitrarily chosen pod from the list.

        Currently the first one added or None if the job has no pods.
        """
        if not self.has_pods():
            return None
        return self.pods[0]

    def get_pods(self) -> List[V1Pod]:
        """Returns the list of job pods.

        Returns in the order of addition to the list or an empty list if no pods.
        """
        return self.pods

    def change_job_name(self, new_name: str):
        """Changes the job name.

        Also the names in its metadata and container specs.
        """
        if self.job.metadata is None:
            self.job.metadata = V1ObjectMeta(name=new_name)
        else:
            self.job.metadata.name = new_name

        if (
            self.job is not None
            and self.job.spec is not None
            and self.job.spec.template is not None
            and self.job.spec.template.metadata is not None
        ):
            self.job.spec.template.metadata.name = new_name

            if self.job.spec.template.spec and self.job.spec.template.spec.containers:
                self.job.spec.template.spec.containers[0].name = new_name

    def get_job_name(self) -> Optional[str]:
        """Returns the job name."""
        return self.job.metadata.name if self.job.metadata else None


class JobStatus(Enum):
    """State of job."""

    ACTIVE = "Active"
    SUCCEEDED = "Succeeded"
    FAILED = "Failed"
